osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -0,0 +1,164 @@
+            GetSecretsResponse response = secretManagementClient.getSecretsRequest(request);
+            Optional<TypedSecret> optionalTypedSecret = getTypedSecretFromSoapResponse(response, SecretType.E_METER_ENCRYPTION_KEY_UNICAST);
+
+            return decryptOptionalSoapSecret(deviceIdentification, optionalTypedSecret);
+        }
+        catch(Exception e) {
+            LOGGER.error("Error while retrieving encryption key", e);
+        }
+        return new byte[0];
+    }
+
+    @Override
+    public byte[] getMbusDefaultKey(String mbusDeviceIdentification) {
+        throw new NotImplementedException();
+    }
+
+    @Override
+    public byte[] getMbusUserKey(String mbusDeviceIdentification) {
+        throw new NotImplementedException();
+    }
+
+    @Override
+    public byte[] getDlmsPassword(String deviceIdentification) {
+        throw new NotImplementedException();
+    }
+
+    @Override
+    public DlmsDevice storeNewKey(DlmsDevice device, byte[] encryptedKey, SecurityKeyType keyType) {
+        throw new NotImplementedException();
+    }
+
+    @Override
+    public DlmsDevice validateNewKey(DlmsDevice device, SecurityKeyType keyType) throws ProtocolAdapterException {
+        throw new NotImplementedException();
+    }
+
+    @Override
+    public byte[] generateKey() {
+        throw new NotImplementedException();
+    }
+
+    @Override
+    public byte[] generateAndEncryptKey() {
+        throw new NotImplementedException();
+    }
+
+    @Override
+    public byte[] encryptMbusUserKey(byte[] mbusDefaultKey, byte[] mbusUserKey) throws ProtocolAdapterException {
+        throw new NotImplementedException();
+    }
+
+    private Optional<TypedSecret> getTypedSecretFromSoapResponse(GetSecretsResponse response, SecretType secretType) {
+        TypedSecrets typedSecrets = response.getTypedSecrets();
+        List<TypedSecret> typedSecretList = typedSecrets.getTypedSecret();
+
+        return typedSecretList.stream().filter(typedSecret -> typedSecret.getType() == SecretType.E_METER_AUTHENTICATION_KEY).findFirst();
+    }
+
+    private GetSecretsRequest getSoapRequestForKey(String deviceIdentification, SecretType secretType) {
+        GetSecretsRequest request = new GetSecretsRequest();
+        request.setDeviceId(deviceIdentification);
+        request.setSecretTypes(new SecretTypes());
+        List<SecretType> secretTypeList = request.getSecretTypes().getSecretType();
+        secretTypeList.add(secretType);
+        return request;
+    }
+
+    private byte[] decryptOptionalSoapSecret(String deviceIdentification, Optional<TypedSecret>optionalTypedSecret) {
+
+        if (optionalTypedSecret.isPresent()) {
+            try {
+                byte[] encryptedSoapSecret = Hex.decodeHex(optionalTypedSecret.get().getSecret());
+                return soapSecretEncryptionService.decrypt(encryptedSoapSecret);
+            }
+            catch (Exception e) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1166)